lite/examples/sound_classification/android/app/src/main/java/org/tensorflow/lite/examples/soundclassifier/MainActivity.kt

```diff
@@ -18,84 +18,151 @@ package org.tensorflow.lite.examples.soundclassifier
+      classificationIntervalSlider.setLabelFormatter { value: Float ->
+        "${value.toInt()} ms"
       }
-    }
-
-    soundClassifier.probabilities.observe(this) { resultMap ->
-      if (resultMap.isEmpty() || resultMap.size > soundClassifier.labelList.size) {
-        Log.w(TAG, "Invalid size of probability output! (size: ${resultMap.size})")
-        return@observe
+      classificationIntervalSlider.addOnChangeListener { _, value, _ ->
+        classificationInterval = value.toLong()
+        stopAudioClassification()
+        startAudioClassification()
       }
-      probabilitiesAdapter.probabilityMap = resultMap
-      probabilitiesAdapter.notifyDataSetChanged()
     }
 
+    // Create a handler to run classification in a background thread
+    val handlerThread = HandlerThread("backgroundThread")
+    handlerThread.start()
+    handler = HandlerCompat.createAsync(handlerThread.looper)
+
+    // Request microphone permission and start running classification
     if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
       requestMicrophonePermission()
     } else {
-      soundClassifier.start()
+      startAudioClassification()
     }
+
+  }
+
+  private fun startAudioClassification() {
+    // If the audio classifier is initialized and running, do nothing.
+    if (audioClassifier != null) return;
+
+    // Initialize the audio classifier
+    val classifier = AudioClassifier.createFromFile(this, MODEL_FILE)
+    val audioTensor = classifier.createInputTensorAudio()
+
+    // Initialize the audio recorder
+    val recorder = classifier.createAudioRecord()
+    recorder.startRecording()
+
+    // Define the classification process
+    val run = object : Runnable {
+      override fun run() {
+        val startTime = System.currentTimeMillis()
+
+        // Load the latest audio sample
+        audioTensor.load(recorder)
+        val output = classifier.classify(audioTensor)
+
+        // Filter out results above a certain threshold, and sort them descendingly
+        val filteredModelOutput = output[0].categories.filter {
+          it.score > MINIMUM_DISPLAY_THRESHOLD
+        }.sortedBy {
+          -it.score
+        }
+
+        val finishTime = System.currentTimeMillis()
+
+        Log.d(TAG, "Latency = ${finishTime - startTime}ms")
+
+        // Updating the UI
+        runOnUiThread {
+          probabilitiesAdapter.categoryList = filteredModelOutput
+          probabilitiesAdapter.notifyDataSetChanged()
+        }
+
+        // Rerun the classification after a certain interval
+        handler.postDelayed(this, classificationInterval)
+      }
+    }
+
+    // Start the classification process
+    handler.post(run)
+
+    // Save the instances we just created for use later
+    audioClassifier = classifier
+    audioRecord = recorder
+  }
+
+  private fun stopAudioClassification() {
+    handler.removeCallbacksAndMessages(null)
+    audioRecord?.stop()
```

Comment:
Is it safe to call this when the audioRecord object might be used concurrently on the Handler thread? I don't think removeCallbacksAndMessages blocks.

Reply:
Even after audioRecord stop recording, it only stop adding new data to its buffer while keeping old data on its buffer. Hence if the classification runnable is running, the TFLite AudioClassifier will classify the old data for one last time before being released.

Setting audioRecord = null doesn't affect the runnable either because it keep the reference to the local `record` instance rather than the activity's `audioRecord`. After the runnable stops, GC will collect the runnable, then the local `record` variable. I don't think there's any concern about memory leaks.